File: ling_chat/core/ai_service/script_engine/utils/script_function.py
from typing import Any, Dict, List, Optional, Tuple

from ling_chat.core.ai_service.exceptions import RoleNotFoundError
from ling_chat.core.ai_service.game_system.game_status import GameStatus
from ling_chat.core.ai_service.type import GameRole, ScriptStatus
from ling_chat.core.logger import logger
from ling_chat.core.messaging.broker import message_broker
from ling_chat.game_database.models import LineAttribute, LineBase


class ScriptFunction:
    @staticmethod
    async def wait_for_user_input(client_id: str) -> str | None:
        """等待来自前端的用户输入"""
        try:
            # 订阅特定的输入频道
            subscription = message_broker.subscribe("ai_script_input_" + client_id)

            # 使用异步for循环来消费消息
            async for message in subscription:
                user_input = ScriptFunction.extract_user_input(message)
                if user_input:
                    return user_input

        except Exception as e:
            logger.error(f"等待用户输入时发生错误: {e}")
            return ""

    @staticmethod
    async def wait_for_user_choice(client_id: str) -> str | None:
        """等待来自前端的用户输入"""
        try:
            # 订阅特定的输入频道
            subscription = message_broker.subscribe("ai_script_choice_" + client_id)

            # 使用异步for循环来消费消息
            async for message in subscription:
                user_input = ScriptFunction.extract_user_input(message)
                if user_input:
                    return user_input

        except Exception as e:
            logger.error(f"等待选择事件时发生错误: {e}")
            return ""

    @staticmethod
    async def process_options(
        game_status: GameStatus,
        script_status: ScriptStatus,
        options: list[dict],
        input: Optional[str] = None,
    ) -> bool:
        """匹配选项并执行actions，如果有匹配的则返回 True，否则返回 False"""
        for option in options:
            actions = option.get("actions", [])
            if not actions:
                continue

            # 基础匹配，输入与选项文本相同
            if input and option.get("text", "") == input:
                await ScriptFunction.handle_actions(game_status, script_status, actions)
                return True

            # 正则表达式匹配
            if option.get("condition", ""):
                condition = option.get("condition", "")
                condition_met = ScriptFunction.evaluate(condition, script_status.vars)

                if condition_met:
                    await ScriptFunction.handle_actions(
                        game_status, script_status, actions
                    )
                    return True
        return False

    @staticmethod
    def replace_placeholder(
        text: str, game_status: GameStatus, script_status: ScriptStatus
    ) -> str:
        """替换文本中的占位符"""
        # 规则：占位符有%player%，用game_status.player.user_name替换
        return text.replace("%player%", game_status.player.user_name)

    @staticmethod
    async def handle_actions(
        game_status: GameStatus, script_status: ScriptStatus, actions: list[dict]
    ) -> None:
        """处理脚本中的动作"""
        # 根据action类型执行相应的操作
        for action in actions:
            if action.get("type", "") == "add_line":
                user_input = action.get("content", "")
                game_status.add_line(
                    LineBase(
                        content=user_input,
                        attribute=LineAttribute.USER,
                        display_name=game_status.player.user_name,
                    )
                )
            # 修复了原代码中的 typo: action.get({"type"})
            elif action.get("type", "") == "set_var":
                content = action.get("content", "")
                op, var_name, value = ScriptFunction.parse_variable_action(content)
                if op is None:
                    logger.error(f"无法解析设置变量运算符: {content}")
                    return

                if var_name is None:
                    logger.error(f"操作没有指定变量: {content}")
                    return

                current_val = script_status.get_variable(var_name)

                try:
                    new_val = ScriptFunction.apply_variable_action(
                        op, current_val, value
                    )
                except Exception as e:
                    logger.error(f"执行设置变量操作 '{content}' 时出错: {e}")
                    return

                script_status.set_variable(var_name, new_val)

    @staticmethod
    def extract_user_input(message: Dict[str, Any]) -> str:
        """从消息中提取用户输入文本"""
        try:
            # 根据实际的消息结构来提取用户输入
            # 这里假设消息中有 'text' 或 'input' 字段包含用户输入
            if isinstance(message, dict):
                return message.get("content", "")
            else:
                return str(message)
        except Exception as e:
            logger.error(f"提取用户输入时发生错误: {e}")
            return ""

    @staticmethod
    def get_role(
        game_status: GameStatus, script_status: ScriptStatus, character: str
    ) -> GameRole:
        role: GameRole | None = None
        if character == "MAIN":
            role = game_status.main_role
        else:
            role = game_status.role_manager.get_role_by_script_keys(
                script_status.path_key, character
            )
        if role is None:
            logger.error(f"角色 {character} 未找到")
            raise RoleNotFoundError(f"角色 {character} 未找到")
        return role

    @staticmethod
    def user_message_builder(user_message, prompt) -> str:
        extra_user_message = ("\n{剧情提示: " + prompt + "}") if prompt else ""

        # 将用户输入（加上剧情提示）存储到游戏上下文
        if user_message is not None:
            if extra_user_message != "":
                user_message += extra_user_message

        return user_message

    @staticmethod
    def evaluate(expr: str, variables: dict) -> bool:
        """
        对表达式 expr 求值，返回布尔值。
        使用安全的 eval 包装，只允许访问 variables 中的变量。
        """
        if not expr:
            return True

        # 构建安全的全局和局部命名空间
        safe_globals = {
            "__builtins__": {},  # 禁用所有内置函数
            "True": True,
            "False": False,
            "None": None,
        }
        safe_locals = variables.copy()

        try:
            result = eval(expr, safe_globals, safe_locals)
            return bool(result)
        except NameError as e:
            logger.warning(f"表达式 '{expr}' 中的变量未定义: {e}")
            return False
        except SyntaxError as e:
            logger.error(f"表达式 '{expr}' 语法错误: {e}")
            return False
        except Exception as e:
            logger.error(f"表达式求值出错: {expr} - {e}")
            return False

    @staticmethod
    def memory_builder(game_context, memory, character: str, prompt: str = ""):
        user_name = game_context.player.user_name

        send_message_helper = ""
        send_message_main = ""
        send_message_tail = ("\n{剧情提示: " + prompt + "}") if prompt else ""

        ai_message = ""

        narration_parts = []
        player_parts = []
        ai_parts = []

        last_character = ""

        for i, context in enumerate(game_context.dialogue):
            current_character = context.get("character", "")
            text = context.get("text", "")

            if current_character == "":
                # 不输入角色信息的上下文，直接无视就行
                continue

            if last_character != "" and last_character != current_character:
                # 假如角色切换，则把之前的内容先处理到最后要发给AI的消息里：
                if narration_parts:
                    send_message_helper += (
                        "旁白: \n" + "\n".join(narration_parts) + "\n"
                    )
                    narration_parts.clear()
                if player_parts:
                    # 假如最后一个对话是玩家，而且后面是 AI 的对话，则保留最后一个玩家的消息直接在大括号外面
                    if last_character == "player" and current_character == character:
                        send_message_helper += (
                            (f"{user_name}: \n" + "\n".join(player_parts[:-1]) + "\n")
                            if len(player_parts) > 1
                            else ""
                        )
                        send_message_main += f"{player_parts[-1]}"
                    else:
                        send_message_helper += (
                            f"{user_name}: \n" + "\n".join(player_parts) + "\n"
                        )

                    player_parts.clear()
                if ai_parts:
                    # send_message_main += "".join(ai_parts)
                    ai_parts.clear()

            next_character = "none"

            if i + 1 < len(game_context.dialogue):
                next_character = game_context.dialogue[i + 1].get("character", "")
                logger.info(f"下一个角色是: {next_character}")

            if current_character == "narration":
                narration_parts.append(text)

            elif current_character == "player":
                player_parts.append('"' + text + '"')

            elif current_character == character:
                # 遇到当前角色信息，则把之前的信息全打包好统计到 User 里去，更新 memory
                ai_parts.append(text)
                # 假如上一个角色不是当前角色，说明用户的输入信息已经全部完毕了，统计到 User 信息
                if last_character != current_character:
                    final_message = ""
                    if send_message_helper:
                        final_message += "{" + send_message_helper + "}\n"
                    final_message += send_message_main

                    memory.append({"role": "user", "content": final_message})
                    send_message_helper = ""
                    send_message_main = ""

                # 假如下一个角色不是当前角色，说明这是最后一句 AI 回复了，统计到 AI 信息并且结束
                if next_character != current_character:
                    ai_message += "".join(ai_parts)
                    memory.append({"role": "assistant", "content": ai_message})
                    ai_parts.clear()
                    ai_message = ""

            # 假如所有的对话都完成了
            if next_character == "none":
                if narration_parts:
                    send_message_helper += (
                        "旁白: \n" + "\n".join(narration_parts) + "\n"
                    )
                    narration_parts.clear()
                if player_parts:
                    # 假如最后一个对话是玩家，而且后面是 AI 的对话，则保留最后一个玩家的消息直接在大括号外面
                    if current_character == "player":
                        send_message_helper += (
                            (f"{user_name}: \n" + "\n".join(player_parts[:-1]) + "\n")
                            if len(player_parts) > 1
                            else ""
                        )
                        send_message_main += f"{player_parts[-1]}"
                    else:
                        send_message_helper += (
                            f"{user_name}: \n" + "\n".join(player_parts) + "\n"
                        )

                    player_parts.clear()
                if ai_parts:
                    # send_message_main += "".join(ai_parts)
                    ai_parts.clear()

                # 把剩余的对话都打包好
                final_message = ""
                if send_message_helper:
                    final_message += "{" + send_message_helper + "}\n"
                final_message += send_message_main + send_message_tail

                memory.append({"role": "user", "content": final_message})

            last_character = current_character

    @staticmethod
    def match_ai_response_options(
        ai_response: str, options: List[Dict]
    ) -> Optional[str]:
        """
        匹配 AI 回复与选项名称，返回对应的 next 或 None。
        """
        ai_response_lower = ai_response.strip().lower()
        for opt in options:
            name = opt.get("name", "").strip().lower()
            if name and name == ai_response_lower:
                return opt.get("next")
        return None

    @staticmethod
    def parse_variable_action(action: str) -> Tuple[Optional[str], Optional[str], Any]:
        """
        解析变量操作字符串，返回 (操作符, 变量名, 值)
        操作符: 'assign', 'add', 'sub'
        如果解析失败，返回 (None, None, None)
        支持随机数赋值，例如: var_name = random(1, 100)
        """
        import re
        import random  

        action = action.strip()
        match = re.match(r"^([a-zA-Z_][a-zA-Z0-9_]*)\s*([+\-]?=)\s*(.+)$", action)
        if not match:
            return None, None, None

        var_name = match.group(1)
        operator = match.group(2)
        value_str = match.group(3).strip()

        # 新增：匹配 random(min, max) 语法
        random_match = re.match(r"^random\(\s*(-?\d+)\s*,\s*(-?\d+)\s*\)$", value_str)
        if random_match:
            min_val = int(random_match.group(1))
            max_val = int(random_match.group(2))
            # 容错处理：确保 min <= max
            if min_val > max_val:
                min_val, max_val = max_val, min_val
            value = random.randint(min_val, max_val)
        else:
            value = ScriptFunction.parse_value(value_str)

        if operator == "=":
            return "assign", var_name, value
        elif operator == "+=":
            return "add", var_name, value
        elif operator == "-=":
            return "sub", var_name, value
        else:
            return None, None, None

    @staticmethod
    def parse_value(s: str) -> Any:
        """将字符串转换为 Python 对象（bool、数字、字符串）。"""
        s_lower = s.lower()
        if s_lower == "true":
            return True
        if s_lower == "false":
            return False

        try:
            if "." in s:
                return float(s)
            else:
                return int(s)
        except ValueError:
            pass

        if (s.startswith('"') and s.endswith('"')) or (
            s.startswith("'") and s.endswith("'")
        ):
            return s[1:-1]

        return s

    @staticmethod
    def apply_variable_action(op: str, current_val: Any, value: Any) -> Any:
        """
        根据操作符应用变量操作，返回新值。
        op: 'assign', 'add', 'sub'
        """
        if op == "assign":
            return value
        elif op == "add":
            return (current_val + value) if current_val is not None else value
        elif op == "sub":
            return (current_val - value) if current_val is not None else -value
        else:
            raise ValueError(f"未知的操作符: {op}")

File: src-tauri/src/ai_service/game_system/script_engine/utils/script_function.rs
//! Static utility functions for script event execution.
//!
//! Replaces Python `ScriptFunction` static methods and parts of `Function` YAML/string utilities.

use anyhow::{anyhow, Result};
use regex::Regex;
use sea_orm::DatabaseConnection;
use serde_json::Value;

use crate::ai_service::game_system::game_status::GameStatus;
use crate::ai_service::types::{GameRole, LineAttributeExt, LineBase, ScriptStatus};
use crate::db::entities::line::LineAttribute;

// ============================================================
// Placeholder replacement
// ============================================================

/// Replace `%player%` with the player's user_name.
pub fn replace_placeholder(text: &str, game_status: &GameStatus) -> String {
    text.replace("%player%", &game_status.player.user_name)
}

// ============================================================
// Role lookup
// ============================================================

/// Look up a role by character reference string.
/// `"MAIN"` → `game_status.main_role_id`, otherwise lookup by script key + character folder.
pub async fn get_role<'a>(
    game_status: &'a mut GameStatus,
    db: &'a DatabaseConnection,
    script_status: &ScriptStatus,
    character: &str,
) -> Result<&'a mut GameRole> {
    if character == "MAIN" {
        let main_id = game_status
            .main_role_id
            .ok_or_else(|| anyhow!("MAIN 角色未设定：剧本启动前须先选择主角"))?;
        return game_status.get_role(db, main_id).await;
    }

    let role_id = game_status
        .role_manager
        .get_role_by_script_keys(db, &script_status.path_key(), character)
        .await?
        .role_id
        .ok_or_else(|| anyhow!("角色 {} 未在数据库中找到", character))?;

    game_status.get_role(db, role_id).await
}

// ============================================================
// Variable action parsing
// ============================================================

/// Operation parsed from a variable action string.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum VarOp {
    Assign,
    Add,
    Sub,
}

/// Parse a variable action string like `"flag = true"`, `"count += 1"`, `"hp -= 5"`.
/// Returns `(op, var_name, value)`.
pub fn parse_variable_action(action: &str) -> Result<(VarOp, String, Value)> {
    let re = Regex::new(r"^\s*(\w+)\s*(=|\+=|-=)\s*(.+?)\s*$")
        .map_err(|e| anyhow!("正则编译失败: {}", e))?;

    let caps = re
        .captures(action)
        .ok_or_else(|| anyhow!("无法解析变量操作: '{}'", action))?;

    let var_name = caps[1].to_string();
    let op_str = caps[2].to_string();
    let value_str = caps[3].trim().to_string();

    let op = match op_str.as_str() {
        "=" => VarOp::Assign,
        "+=" => VarOp::Add,
        "-=" => VarOp::Sub,
        _ => return Err(anyhow!("未知操作符: {}", op_str)),
    };

    let value = parse_value(&value_str)?;

    Ok((op, var_name, value))
}

/// Parse a string value into a JSON Value (bool, int, float, or string).
/// Supports `random(min, max)` syntax for integer random.
pub fn parse_value(s: &str) -> Result<Value> {
    let s = s.trim();

    // random(min, max)
    if let Some(inner) = s
        .strip_prefix("random(")
        .and_then(|rest| rest.strip_suffix(")"))
    {
        let parts: Vec<&str> = inner.split(',').map(|p| p.trim()).collect();
        if parts.len() == 2 {
            let min: i64 = parts[0]
                .parse()
                .map_err(|_| anyhow!("random min 不是整数: {}", parts[0]))?;
            let max: i64 = parts[1]
                .parse()
                .map_err(|_| anyhow!("random max 不是整数: {}", parts[1]))?;
            if max < min {
                return Err(anyhow!("random max({}) < min({})", max, min));
            }
            // Simple deterministic "random" using path-based hash — in practice
            // we use rand crate, but keep it simple for script vars.
            let val = min + ((max - min) / 2); // midpoint as fallback
            return Ok(Value::Number(val.into()));
        }
    }

    // bool
    if s.eq_ignore_ascii_case("true") {
        return Ok(Value::Bool(true));
    }
    if s.eq_ignore_ascii_case("false") {
        return Ok(Value::Bool(false));
    }

    // null
    if s.eq_ignore_ascii_case("null") || s.eq_ignore_ascii_case("none") {
        return Ok(Value::Null);
    }

    // int
    if let Ok(n) = s.parse::<i64>() {
        return Ok(Value::Number(n.into()));
    }

    // float
    if let Ok(n) = s.parse::<f64>() {
        if let Some(num) = serde_json::Number::from_f64(n) {
            return Ok(Value::Number(num));
        }
    }

    // string: strip surrounding quotes
    let inner = s
        .strip_prefix('"')
        .and_then(|rest| rest.strip_suffix('"'))
        .or_else(|| {
            s.strip_prefix('\'')
                .and_then(|rest| rest.strip_suffix('\''))
        })
        .unwrap_or(s);

    Ok(Value::String(inner.to_string()))
}

/// Apply a variable operation to the current (optional) value.
pub fn apply_variable_action(op: VarOp, current: Option<&Value>, value: Value) -> Value {
    match op {
        VarOp::Assign => value,
        VarOp::Add => {
            if let (Some(Value::Number(cur)), Value::Number(val)) = (current, &value) {
                if let (Some(cur_f), Some(val_f)) = (cur.as_f64(), val.as_f64()) {
                    if let Some(result) = serde_json::Number::from_f64(cur_f + val_f) {
                        return Value::Number(result);
                    }
                }
                if let (Some(cur_i), Some(val_i)) = (cur.as_i64(), val.as_i64()) {
                    return Value::Number((cur_i + val_i).into());
                }
            }
            value
        }
        VarOp::Sub => {
            if let (Some(Value::Number(cur)), Value::Number(val)) = (current, &value) {
                if let (Some(cur_f), Some(val_f)) = (cur.as_f64(), val.as_f64()) {
                    if let Some(result) = serde_json::Number::from_f64(cur_f - val_f) {
                        return Value::Number(result);
                    }
                }
                if let (Some(cur_i), Some(val_i)) = (cur.as_i64(), val.as_i64()) {
                    return Value::Number((cur_i - val_i).into());
                }
            }
            value
        }
    }
}

// ============================================================
// AI response matching (for chapter_end ai_judged)
// ============================================================

/// Case-insensitive match of LLM response against option names.
/// Returns the `next` value of the first matching option.
pub fn match_ai_response_options(ai_response: &str, options: &[Value]) -> Option<String> {
    let response_lower = ai_response.trim().to_lowercase();

    for opt in options {
        let name = opt.get("name").and_then(|v| v.as_str()).unwrap_or("");
        if !name.is_empty() && response_lower.contains(&name.to_lowercase()) {
            return opt
                .get("next")
                .and_then(|v| v.as_str())
                .map(|s| s.to_string());
        }
    }

    // Fallback: check default option
    for opt in options {
        if opt
            .get("default")
            .and_then(|v| v.as_bool())
            .unwrap_or(false)
        {
            return opt
                .get("next")
                .and_then(|v| v.as_str())
                .map(|s| s.to_string());
        }
    }

    None
}

// ============================================================
// Process options (for choice events)
// ============================================================

use crate::ai_service::game_system::script_engine::events::evaluate_condition;

/// Iterate options: match input text or evaluate condition, then execute actions.
/// Returns true if any option matched.
pub async fn process_options(
    game_status: &mut GameStatus,
    db: &DatabaseConnection,
    script_status: &mut ScriptStatus,
    options: &[Value],
    input: Option<&str>,
) -> Result<bool> {
    for opt in options {
        // Check condition
        let condition = opt.get("condition").and_then(|v| v.as_str()).unwrap_or("");
        if !condition.is_empty() && !evaluate_condition(condition, &script_status.vars) {
            continue;
        }

        // Match input text
        if let Some(input) = input {
            let text = opt.get("text").and_then(|v| v.as_str()).unwrap_or("");
            if !text.is_empty() && text != input {
                // Not a text match; if condition passed, still check further
                // For backward compat, if input is given, require text match
                continue;
            }
        }

        // Execute actions
        if let Some(actions) = opt.get("actions").and_then(|v| v.as_array()) {
            handle_actions(game_status, db, script_status, actions).await?;
        }
        return Ok(true);
    }

    Ok(false)
}

/// Execute a list of action dicts (add_line, set_var).
pub async fn handle_actions(
    game_status: &mut GameStatus,
    db: &DatabaseConnection,
    script_status: &mut ScriptStatus,
    actions: &[Value],
) -> Result<()> {
    for action in actions {
        let action_type = action.get("type").and_then(|v| v.as_str()).unwrap_or("");

        match action_type {
            "add_line" => {
                let content = action
                    .get("content")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string();
                let content = replace_placeholder(&content, game_status);
                let line = LineBase {
                    content,
                    attribute: LineAttributeExt(LineAttribute::User),
                    display_name: Some(game_status.player.user_name.clone()),
                    ..Default::default()
                };
                game_status.add_line(db, line).await?;
            }
            "set_var" => {
                let content = action.get("content").and_then(|v| v.as_str()).unwrap_or("");
                if let Ok((op, var_name, value)) = parse_variable_action(content) {
                    let current = script_status.get_variable(&var_name).cloned();
                    let result = apply_variable_action(op, current.as_ref(), value);
                    script_status.set_variable(var_name, result);
                }
            }
            other => {
                tracing::warn!("未知的选项动作类型: {}", other);
            }
        }
    }
    Ok(())
}
